s3fs/core.py

```diff
@@ -1311,6 +1311,49 @@ async def _copy_basic(self, path1, path2, **kwargs):
             raise ValueError("Copy failed (%r -> %r): %s" % (path1, path2, e)) from e
         self.invalidate_cache(path2)
 
+    async def _copy_etag_preserved(self, path1, path2, size, total_parts, **kwargs):
+        """Copy file between locations on S3 as multi-part while preserving
+        the etag (using the same part sizes for each part"""
+
+        bucket1, key1, version1 = self.split_path(path1)
+        bucket2, key2, version2 = self.split_path(path2)
+
+        mpu = await self._call_s3(
+            self.s3.create_multipart_upload, Bucket=bucket2, Key=key2, **kwargs
+        )
+
+        parts = []
+        brange_first = 0
+        for i in range(1, total_parts + 1):
+            part_info = await self._call_s3(
+                self.s3.head_object, Bucket=bucket1, Key=key1, PartNumber=i
```

Comment:
+1

Reply:
We start from range(1, ...) so I assume it is not needed. I would have used +1 if we used `i` only once though we use it 2 times (one for head_object and one for upload part)

Reply:
I mean that as in "I like it" :)